cmd/dra-example-kubeletplugin/state.go

```diff
@@ -122,53 +82,37 @@ func NewDeviceState(config *Config) (*DeviceState, error) {
 	return state, nil
 }
 
-func (s *DeviceState) Prepare(claimUID string, allocation AllocatedDevices) ([]string, error) {
+func (s *DeviceState) Prepare(claim *resourceapi.ResourceClaim) ([]*drapbv1.Device, error) {
 	s.Lock()
 	defer s.Unlock()
 
+	claimUID := string(claim.UID)
+
 	checkpoint := newCheckpoint()
 	if err := s.checkpointManager.GetCheckpoint(DriverPluginCheckpointFile, checkpoint); err != nil {
 		return nil, fmt.Errorf("unable to sync from checkpoint: %v", err)
 	}
 	preparedClaims := checkpoint.V1.PreparedClaims
 
 	if preparedClaims[claimUID] != nil {
-		cdiDevices, err := s.cdi.GetClaimDevices(claimUID, preparedClaims[claimUID])
-		if err != nil {
-			return nil, fmt.Errorf("unable to get CDI devices names: %v", err)
-		}
-		return cdiDevices, nil
+		return preparedClaims[claimUID], nil
 	}
 
-	preparedDevices := &PreparedDevices{}
-
-	var err error
-	switch allocation.Type() {
-	case gpucrd.GpuDeviceType:
-		preparedDevices.Gpu, err = s.prepareGpus(claimUID, allocation.Gpu)
-	default:
-		err = fmt.Errorf("unknown device type: %v", allocation.Type())
-	}
+	preparedDevices, err := s.prepareDevices(claim)
 	if err != nil {
-		return nil, fmt.Errorf("praparation failed: %v", err)
+		return nil, fmt.Errorf("prepare failed: %v", err)
 	}
 
-	err = s.cdi.CreateClaimSpecFile(claimUID, preparedDevices)
-	if err != nil {
+	if err = s.cdi.CreateClaimSpecFile(claimUID, preparedDevices); err != nil {
```

Comment:
What do you mean by "when the same device is used in multiple claims"? Under what circumstances would this happen? In this example driver it will *never* happen because each claim creates a new CDI spec file with a single CDI device in it. That device is named with the claimUID.

Reply:
On the line 163[new code] below - the device name appears to be a device name from allocation result, not a claimUID, permalink: https://github.com/kubernetes-sigs/dra-example-driver/blob/5bb4cfe4bbfef7510cb88556d9e96b6b37541762/cmd/dra-example-kubeletplugin/state.go#L163

That means if the device is selected based on device property, for instance memory, for one claim, and it has enough of same resource for second claim - it might get selected again for another claim, no? If yes - then there will be two CDI spec files with different name, but with same kind and with same device name, which will be a conflict.

I might be wrong, perhaps I just need to build this branch and run it.

Reply:
I guess I didn't do the claimUID --> CDI Device name mapping in this driver.

Regardless of that, I don't quite follow your example. Once a device is allocated by the scheduler it is unavailable for allocation to another claim. The only exception to that is devices allocated in management mode -- in that case you can (and do) get access to devices that are allocated to other claims.

Reply:
All of that said, a better model might be to generate a CDI spec once for all possible devices and then just referencing them as devices get prepared, rather than generating a new spec per claim.

Reply:
Let's do that as a follow-up though since it is a bug fix / feature addition rather than a refactoring as this PR does.

Reply:
I agree.